waspc/data/Generator/templates/react-app/src/actions/index.js

```diff
@@ -0,0 +1,227 @@
+
+  const definitionErrors = []
+  if (typeof getQuerySpecifier !== 'function') {
+    definitionErrors.push('`getQuerySpecifier` is not a function.')
+  }
+  if (typeof updateQuery !== 'function') {
+    definitionErrors.push('`updateQuery` is not a function.')
+  }
+  if (definitionErrors.length) {
+    throw new TypeError(`Invalid optimistic update definition: ${definitionErrors.join(', ')}.`)
+  }
+
+  return {
+    getQueryKey: (item) => getRqQueryKeyFromSpecifier(getQuerySpecifier(item)),
+    updateQuery,
+  }
+}
+
+/**
+ * This function implements the methods necessary for configuring optimistic
+ * updates using React Query, as described by their documentation:
+ * https://tanstack.com/query/v4/docs/guides/optimistic-updates?from=reactQueryV3&original=https://react-query-v3.tanstack.com/guides/optimistic-updates
+ *
+ * @param {Object} queryClient The QueryClient instance used by React Query.
+ * @param {InternalOptimisticUpdateDefinition} optimisticUpdateDefinitions A list containing internal optimistic updates definition objects
+ * (i.e., a list where each object carries the instructions for performing particular optimistic update).
+ * @returns {Object} An object containing 'onMutate' and 'onError' functions corresponding to the given optimistic update
+ * definitions (check React Query's docs for details).
+ */
+function makeRqOptimisticUpdateOptions(queryClient, optimisticUpdateDefinitions) {
+  async function onMutate(item) {
+    const specificOptimisticUpdateDefinitions = optimisticUpdateDefinitions.map(
+      generalDefinition => getOptimisticUpdateDefinitionForSpecificItem(generalDefinition, item)
+    )
+
+    // Cancel any outgoing refetches (so they don't overwrite our optimistic update).
+    // Theoretically, we can be a bit faster. Instead of awaiting the
+    // cancellation of all queries, we could cancel and update them in parallel.
+    // However, awaiting cancellation hasn't yet proven to be a performance bottleneck.
+    await Promise.all(specificOptimisticUpdateDefinitions.map(
+      ({ queryKey }) => queryClient.cancelQueries(queryKey)
+    ))
+
+    // We're using a Map to to correctly serialize query keys that contain objects.
+    const previousData = new Map()
+    specificOptimisticUpdateDefinitions.forEach(({ queryKey, updateQuery }) => {
+      // Snapshot the currently cached value.
+      const previousDataForQuery = queryClient.getQueryData(queryKey)
+
+      // Attempt to optimistically update the cache using the new value.
+      try {
+        queryClient.setQueryData(queryKey, updateQuery)
+      } catch (e) {
+        console.error("The `updateQuery` function threw an exception, skipping optimistic update:")
+        console.error(e)
+      }
+
+      // Remember the snapshotted value to restore in case of an error.
+      previousData.set(queryKey, previousDataForQuery)
+    })
+
+    return { previousData }
+  }
+
+  function onError(err, item, context) {
```

Comment:
Super small, but you could add underscore to `err` and `item` due to not being used.

Reply:
Ah yes, I wanted to but forgot. Thanks!